usl_pipeline/cloud_functions/main.py

```diff
@@ -903,3 +900,98 @@ def _write_as_npy(blob: storage.Blob, array: NDArray) -> None:
     npy_file.flush()
     npy_file.seek(0)
     blob.upload_from_file(npy_file)
+
+
+def _start_feature_rescaling_if_ready(feature_bucket: storage.Bucket, blob_path: str):
+    """Creates scaled version for every chunk file in the storage bucket parent folder.
+
+    Output files with scaled feature matrices will be placed to the same folder where
+    unscaled files (following "chunk_*" name pattern) are stored but will get "scaled_"
+    prefix. Rescaling only starts when metadata of study area (corresponding to the
+    parent folder name) contains chunks counters giving exactly the same number of
+    chunks as are currently stored in the bucket folder. This should happen when the
+    last feature chunk is stored.
+
+    Args:
+        feature_bucket: Storage bucket with feature matrices.
+        blob_path: Path pointing to one of unscaled feature matrix files with the name
+            following the "chunk_*" pattern. Other files are ignored.
+    """
+    study_area_name, chunk_name = _parse_chunk_path(blob_path)
+    # Let's look up chunk metadata and check if scaling is needed
+    db = firestore.Client()
+    chunk_metadata = metastore.StudyAreaSpatialChunk.get_if_exists(
+        db, study_area_name, chunk_name
+    )
+    if chunk_metadata is None:
+        logging.info(f"Chunk metadata was not found for {study_area_name}/{chunk_name}")
+        return
+
+    if not chunk_metadata.needs_scaling:
+        logging.info(f"Chunk {study_area_name}/{chunk_name} doesn't need scaling")
+        return
+
+    study_area = metastore.StudyArea.get(db, study_area_name)
+    if study_area.chunk_x_count is None or study_area.chunk_y_count is None:
+        logging.info(
+            f"[Feature Rescaler] Study area {study_area_name} has no chunk counts in "
+            + "metadata"
+        )
+        return
+
+    feature_blobs = list(feature_bucket.list_blobs(prefix=f"{study_area_name}/chunk_"))
+
+    chunk_count = study_area.chunk_x_count * study_area.chunk_y_count
+    if chunk_count != len(feature_blobs):
+        logging.info(
+            f"[Feature Rescaler] Study area {study_area_name} has chunk counts"
+            + f" in metadata ({chunk_count}) different from number of stored feature"
+            + f" chunks ({len(feature_blobs)})"
+        )
+        return
+
+    # We're at the final step when all unscaled feature matrices are ready
+    for feature_blob in feature_blobs:
+        logging.info(f"[Feature Rescaler] Rescaling feature matrix {feature_blob.name}")
+        _, blob_chunk_name = _parse_chunk_path(feature_blob.name)
+        output_blob = feature_bucket.blob(
+            f"{study_area_name}/scaled_{blob_chunk_name}.npy"
+        )
+        with feature_blob.open("rb") as in_fd, output_blob.open("wb") as out_fd:
```

Comment:
By its nature, this will be a run-at-least-once rather than run-exactly-once operation. We could try to scale the same files twice. We should add similar exception handling as I have here:
https://github.com/UrbanSystemsLab/climateiq-cnn/blob/main/usl_pipeline/cloud_functions/main.py#L503
in case we're running twice and the other invocation has started deleting chunks.

Reply:
Actually, I think it's much safer to trigger scaling processing separately for each chunk. See my recent refactoring related to this.

Reply:
Cloud functions by their nature are run-at-least-once rather than run-exactly-once, so we could still end up with two invocations trying to scale the same one file, so we could potentially still see a deletion while trying to read here. Certainly less likely with your refactoring, though!

Reply:
I added the try-except-not-found check that you proposed above.